src/main/scala/com/microsoft/hyperspace/index/ExperimentalStructure.scala

```diff
@@ -0,0 +1,83 @@
+package com.microsoft.hyperspace.index
+
+import java.io.FileNotFoundException
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileStatus, Path}
+
+import com.microsoft.hyperspace.util.{JsonUtils, PathUtils}
+// scalastyle:off
+case class Content1(root: Directory) {
+  @JsonIgnore
+  lazy val files: Seq[Path] = rec(new Path(root.name), root)
+
+  private def rec(prefixPath: Path, directory: Directory): Seq[Path] = {
+    val files = directory.files.map(f => new Path(prefixPath, f.name))
+    files ++ directory.subDirs.flatMap {
+      dir => rec(new Path(prefixPath, dir.name), dir)
+    }
+  }
+}
+object Content1 {
+  def apply(path: Path): Content1 = Content1(Directory(path))
+}
+
+case class Directory(
+  name: String,
+  files: Seq[FileInfo] = Seq(),
+  subDirs: Seq[Directory] = Seq())
+
+object Directory {
+  def apply(path: Path): Directory = {
+    val files: Seq[FileInfo] = {
+      try {
+        val fs = path.getFileSystem(new Configuration)
+        val statuses = fs.listStatus(path)
+
+        // Assuming index directories don't contain nested directories. Only Leaf files.
+        assert(statuses.forall(!_.isDirectory))
```

Comment:
by `all files` you mean files which represent **this** data or index snapshot, all **all the files in these directories**?

Reply:
I meant that this function should not assume "index" only scenario since this can be used for source as well.

Reply:
I agree. Also, from what I remember, we did agree to support this scenario (where we list everything in a given directory recursively), so maybe let's support it rightaway.

Reply:
thanks @imback82 @rapoth , the recursive structure is supported by the design so _if the user chooses to, they can build a recursive tree_. But in majority of our cases, we will _not list everything recursively_ in the file structure. 
The intention is, we want to support this case: if file 'a/b/f1.parquet` is _not_ added to the `Content` object, then this will not belong to the content, even if the file exists on disk.

Reply:
Here is my suggestion. Either 1) Introduce knobs to support different (general) scenarios or 2) Move out this specific logic to a place requires this behavior.